### ISSUE TITLE
New field in `games.json`: `store_ids`

### ISSUE THREAD

**Issue by felipecrs** [Starz0r/AreWeAntiCheatYet#559]
This point was raised in:

- https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/1504#discussion_r914254657

Where the plan is to use [AreWeAntiCheatYet](https://github.com/Starz0r/AreWeAntiCheatYet) to fetch the required metadata for displaying anti-cheat status directly in Heroic.

There is a problem, however. Some game names does not match and we need to create maps manually, like [this](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/1504/files/ff4ec651155f2b732fafa9bbd63b9a5735e95ef3#diff-32b5c1cee2619eac8a98688432c1657be04475b7c1e8af44a53b5080301ac241).

It would be great if we could instead contribute a new field to `games.json`, like `store_ids`, and it would be ~this:

```json
{
    "url": "",
    "name": "Fall Guys: Ultimate Knockout",
    "logo": "",
    "native": false,
    "status": "Running",
    "reference": "https://www.protondb.com/app/1097150#aRzxJIFk7c",
    "anticheats": [
        "Easy Anti-Cheat"
    ],
    "notes": [
        [
            "Requires workaround",
            "https://steamcommunity.com/sharedfiles/filedetails/?id=2787044978"
        ]
    ],
    "updates": [
        {
            "name": "Enabled Linux Support",
            "date": "Mon 28, March 2022, 8:38:44 PM GMT+02:00",
            "reference": "https://www.protondb.com/app/1097150#aRzxJIFk7c"
        }
    ],
    "store_ids": {
      "steam": "1097150",
      "epic": "0a2d9f6403244d12969e11da6713137b"
    }
}
```

That would allow us to disambiguate games very sharply.

PS: `steam` was added as an example only, we would not use it in Heroic.

PS2: this is not a request for you to maintain the list of store_ids up-to-date (PS: once added, they don't change). It's a request for this new field, and we would update it as needed to disambiguate games in Heroic.

But I do think it can be useful in other scenarios, like to display a link to Steam for example in your website.

So, what do you think?

**Comment by Curve** [comment]
Sounds good!

I will push some changes with which we could also make use of this on the website, for example open the store page

**Comment by Curve** [comment]
I'm done with adding a lot of steam-ids and opening the store page with the appId works great.  
However I can't figure out how to open the store page of an epic game by using its ID.  
Would you happen to know how one can do that?

**Comment by felipecrs** [comment]
The Steam games id are easy to find indeed. They can be found in the URL of the games. Like:

https://store.steampowered.com/app/1097150/Fall_Guys/ -> 1097150

Which can be easily translated to a URL, like:

https://store.steampowered.com/app/1097150/Fall_Guys/

To get the Heroic game id (it's called App Name) is a bit harder though. The easiest way I found was to use Legendary:

```console
$ legendary info "fall guys"
Game Information:
- App name: 0a2d9f6403244d12969e11da6713137b
```

But it does not work all the time. Like for _Battlefield 2042_ which I could not make it work at all. As last resource it's always possible to install the game in your computer, create a shortcut for it through the Epic Games Launcher, and get the id from there:

![image](https://user-images.githubusercontent.com/29582865/177790446-9c7f2099-d8e5-46f4-a46d-3ea1e6438fb5.png)

I'm hopeful that:

- https://github.com/acidicoala/ScreamDB/issues/4

Can be implemented, so that we could make this step easier by using ScreamDB: https://scream-db.web.app/.

Another problem, though, is that it's not easy to translate those _App Names_ to a link to the Epic Games Store.

The Epic Games Store URLs, like:

https://store.epicgames.com/en-US/p/fall-guys

Seems to always use a slugfyied version of the game name as shown in Epic.

@arielj suggested that we could add `store_name` instead of `store_id`. This would allow the Heroic to disambiguate still, and yet would allow you to slugify this name and (most likely) create a link to the EGS page.

But I really wanted to avoid that route as the store names can change over time (devs sometimes decide to rename their games), and therefore could add a maintenance burden to AreWeAntiCheatYet.

It's, however, easy to find:

https://store.epicgames.com/en-US/p/fall-guys

![image](https://user-images.githubusercontent.com/29582865/177792367-e2a4f885-9315-4176-8b8c-cc9f5329caf2.png)

**Comment by Curve** [comment]
https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/1504#discussion_r915934070

TLDR: Why not store both the ID and the StoreName? We could both profit if we do it that way ^^

**Comment by felipecrs** [comment]
I found something by the way that can be useful: https://database.egdata.app/offer/07ce78560aa34180936b199202274462

**Comment by Curve** [comment]
> 

If you think going that way is more robust, sure. I'm happy as long as I can **easily** get a store link to the game from the information stotred

**Comment by felipecrs** [comment]
> TLDR: Why not store both the ID and the StoreName? We could both profit if we do it that way ^^

If we go that route, I suggest you only add the store name. That would be enough for Heroic.

**Comment by Curve** [comment]
> 

Well, we can use whatever suits Heroic the best :)

**Comment by arielj** [comment]
Just to clarify on the Epic store urls, the url is not always a slug generated with the app name (though it is in many cases), we are currently doing something like that in Heroic but it's not correct in many cases (for reference, I opened this issue at some point https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/1160).

I don't think there's a simple way to generate the url from the game's name or the game's id for Epic, I'm not familiar with the EpicData API either.

I'm checking EpicData now and we don't have the ID, we do have the namespace, but I see the URL with the info requires the offer_id and we don't have that information in the game's metadata.

This is the metadata for Batman Arkham Asylum Game of the Year Editor for example:
```
{
  "app_name": "Godwit",
  "app_title": "Batman\u2122 Arkham Asylum Game of the Year Edition",
  "app_version": "1.93",
  "asset_info": {
    "app_name": "Godwit",
    "asset_id": "Godwit",
    "build_version": "1.93",
    "catalog_item_id": "a3361123cbb5436ab8630caea6e1d39a",
    "label_name": "Live-Windows",
    "metadata": {},
    "namespace": "5f7c811695f14b8fa3a8e1ea35713d23"
  },
  "base_urls": [],
  "metadata": {
    "categories": [
      {
        "path": "public"
      },
      {
        "path": "games"
      },
      {
        "path": "applications"
      }
    ],
    "creationDate": "2019-08-27T22:18:50.207Z",
    "customAttributes": {
      "CanRunOffline": {
        "type": "STRING",
        "value": "true"
      },
      "CanSkipKoreanIdVerification": {
        "type": "STRING",
        "value": "true"
      },
      "FolderName": {
        "type": "STRING",
        "value": "BatmanArkhamAsylum"
      },
      "MonitorPresence": {
        "type": "STRING",
        "value": "true"
      },
      "PresenceId": {
        "type": "STRING",
        "value": "5f7c811695f14b8fa3a8e1ea35713d23"
      },
      "RequirementsJson": {
        "type": "STRING",
        "value": "[]"
      },
      "UseAccessControl": {
        "type": "STRING",
        "value": "false"
      }
    },
    "description": "Batman\u2122 Arkham Asylum Game of the Year Edition",
    "developer": "Warner Bros.",
    "developerId": "o-7kyv8uwtz74njnl544kbxzt5xqephw",
    "endOfSupport": false,
    "entitlementName": "a3361123cbb5436ab8630caea6e1d39a",
    "entitlementType": "EXECUTABLE",
    "eulaIds": [
      "egstore"
    ],
    "id": "a3361123cbb5436ab8630caea6e1d39a",
    "itemType": "DURABLE",
    "keyImages": [
      {
        "height": 100,
        "md5": "96dd802d2adb46614b82db64a6b72abf",
        "size": 22781,
        "type": "DieselGameBoxLogo",
        "uploadedDate": "2019-09-16T21:25:20.208Z",
        "url": "https://cdn1.epicgames.com/5f7c811695f14b8fa3a8e1ea35713d23/item/batman-arkham-asylum-209x100-96dd802d2adb46614b82db64a6b72abf.png",
        "width": 209
      },
      {
        "height": 866,
        "md5": "f145139d16a584ae891b77bb0300e373",
        "size": 398134,
        "type": "DieselGameBox",
        "uploadedDate": "2019-09-16T21:26:02.283Z",
        "url": "https://cdn1.epicgames.com/5f7c811695f14b8fa3a8e1ea35713d23/item/batman-arkham-asylum_promo-1704x866-f145139d16a584ae891b77bb0300e373.jpg",
        "width": 1704
      },
      {
        "height": 1267,
        "md5": "5d89c5e6b96ba2db5e607521b2d0ae3e",
        "size": 401333,
        "type": "DieselGameBoxTall",
        "uploadedDate": "2019-10-22T09:41:44.790Z",
        "url": "https://cdn1.epicgames.com/5f7c811695f14b8fa3a8e1ea35713d23/item/batman-arkham-asylum_portrait-1093x1267-5d89c5e6b96ba2db5e607521b2d0ae3e.jpg",
        "width": 1093
      }
    ],
    "lastModifiedDate": "2020-01-07T18:14:40.392Z",
    "namespace": "5f7c811695f14b8fa3a8e1ea35713d23",
    "releaseInfo": [
      {
        "appId": "Godwit",
        "dateAdded": "2019-08-27T00:00:00.000Z",
        "id": "45e4c4db530a422f931338f6d5b3c254",
        "platform": [
          "Windows"
        ]
      }
    ],
    "selfRefundable": true,
    "status": "ACTIVE",
    "title": "Batman\u2122 Arkham Asylum Game of the Year Edition",
    "unsearchable": false
  }
}
```

**Comment by felipecrs** [comment]
@arielj EpicData can search by namespace:

![image](https://user-images.githubusercontent.com/29582865/177831479-a661f3bf-b4be-4441-8af7-8a6f7391d513.png)

**Comment by Curve** [comment]
Looks good! However the Repo hasn't updated in 2 years, so I'm not sure if it's a viable option

**Comment by felipecrs** [comment]
EpicData isn't being updated indeed, but this "fork" is, somehow.

This Discord channel shows the messages of a bot updating the database and it happens frequently:

https://discord.com/channels/561598657879605269/798962119180943400

**Comment by felipecrs** [comment]
https://github.com/nachoaldamav/offers-tracker

**Comment by felipecrs** [comment]
@arielj probably it can also be used to fix your Heroic issue:

- https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/1160

**Comment by felipecrs** [comment]
So here is my suggestion:

- Add both Epic `namespace` and `slug` to AWACY.
- Both can be easily found in the https://database.egdata.app/offers/ by searching:

![image](https://user-images.githubusercontent.com/29582865/177835419-7516f55a-fcd3-484b-9ded-df05ead4854a.png)

- `namespace` can be used by Heroic (@arielj please confirm if you'd be ok)
- `slug` can be used by AWACY to create the link to EGS. Just ensure to remove any `/home` suffix from it.

What do you guys think?

**Comment by felipecrs** [comment]
Unless... @arielj wants to make Heroic aware of the slug by fetching it using the [method described by him](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/issues/1160#issuecomment-1080812578) and therefore only `slug` would be required to be added to AWACY. :-P

**Comment by arielj** [comment]
I only checked one game and the namespace value we get on the metadata matches the one in EpicData, so we could use that value to match the game in the AWACY games.json. I'm assuming the rest of the games are correct.

**Comment by felipecrs** [comment]
Do you know a way to check the metadata of games from within Heroic? So I could double check some games.

**Comment by arielj** [comment]
Hmmm I don't think we use the namespace anywhere in heroic, but you can go to `~/.config/Legendary/metadata` and you'll see all the json files with the metadata of the games you own.

**Comment by felipecrs** [comment]
Awesome, thank you. I'll check a few here.

**Comment by Curve** [comment]
I went ahead and added the `namespace` and `slug` properties to the games list.  
If there's anything you'd like to have changed please let me know and I'll adjust it asap

**Comment by felipecrs** [comment]
Thanks a lot, @Curve. You are amazing!
